File: 2429-design-a-food-rating-system/design-a-food-rating-system.java
class Food{
    String name, cuisine;
    int rating;
    Food() {}
    Food(String name, String cuisine, int rating) {
        this.name = name;
        this.cuisine = cuisine;
        this.rating = rating;
    }
}

class FoodRatings {

    Map<String, Food> foodMap;
    Map<String, TreeSet<Food>> cuisineMap;
    Comparator<Food> foodComparator = (f1, f2) -> {
        if (f1.rating == f2.rating) {
            return f1.name.compareTo(f2.name); // lexicographically smaller first
        }
        return Integer.compare(f2.rating, f1.rating); // higher rating first
    };

    public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
        foodMap = new HashMap<>();
        cuisineMap = new HashMap<>();
        for(int i = 0; i < foods.length; i++) {
            Food food = new Food(foods[i], cuisines[i], ratings[i]);
            foodMap.put(food.name, food);
            cuisineMap.putIfAbsent(cuisines[i], new TreeSet<>(foodComparator));
            cuisineMap.get(cuisines[i]).add(food);
        }
    }

    public void changeRating(String food, int newRating) {
        if(!foodMap.containsKey(food)) {
            return;
        }
        
        // get hold of the old-food
        Food oldFood = foodMap.get(food);
        //remove it from the cuisineMap
        cuisineMap.get(oldFood.cuisine).remove(oldFood);
        
        //change the rating
        oldFood.rating = newRating;
        //put the updated food in cuisineMap
        cuisineMap.get(oldFood.cuisine).add(oldFood);
    }

    public String highestRated(String cuisine) {
        return cuisineMap.get(cuisine).getFirst().name;
    }
}

File: 2429-design-a-food-rating-system/design-a-food-rating-system.cpp
struct Food {
    string name, cuisine;
    int rating;
};

struct FoodComparator {
    bool operator()(const Food* f1, const Food* f2) const {
        if (f1->rating == f2->rating) {
            return f1->name < f2->name;  // lexicographically smaller comes first
        }
        return f1->rating > f2->rating; // higher rating first
    }
};

class FoodRatings {
    unordered_map<string, Food*> foodMap;  // food -> Food*
    unordered_map<string, set<Food*, FoodComparator>> cuisineMap; // cuisine -> ordered set

public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        for (int i = 0; i < foods.size(); i++) {
            Food* food = new Food{foods[i], cuisines[i], ratings[i]};
            foodMap[foods[i]] = food;
            cuisineMap[cuisines[i]].insert(food);
        }
    }

    void changeRating(string food, int newRating) {
        if (!foodMap.count(food)) return;
        Food* f = foodMap[food];
        cuisineMap[f->cuisine].erase(f); // remove old
        f->rating = newRating;
        cuisineMap[f->cuisine].insert(f); // reinsert
    }

    string highestRated(string cuisine) {
        return (*cuisineMap[cuisine].begin())->name; // first element = best
    }
};
